File: src/harness/batch/BatchProcessor.ts
/**
 * 批处理引擎
 *
 * 并行运行多个 prompt，生成结构化轨迹数据
 * 设计参考: Hermes Agent 批处理系统
 */

import { Logger } from '../../utils/Logger';

export interface BatchConfig {
  /** 最大并发数 */
  concurrency: number;
  /** 单个任务超时（毫秒） */
  timeout: number;
  /** 输出格式 */
  outputFormat?: 'sharegpt' | 'jsonl' | 'raw';
  /** 部分失败时是否继续 */
  continueOnError?: boolean;
}

export interface BatchPrompt {
  /** 唯一标识 */
  id: string;
  /** 提示文本 */
  text: string;
  /** 系统提示（可选） */
  systemPrompt?: string;
  /** 元数据 */
  metadata?: Record<string, unknown>;
}

export interface BatchItemResult {
  /** 对应 prompt 的 ID */
  id: string;
  /** 响应文本 */
  response: string;
  /** 是否成功 */
  success: boolean;
  /** 执行耗时（毫秒） */
  duration: number;
  /** 错误信息 */
  error?: string;
  /** 元数据 */
  metadata?: Record<string, unknown>;
}

export interface ShareGPTConversation {
  conversations: Array<{ from: 'human' | 'gpt'; value: string }>;
}

export class BatchProcessor {
  private config: Required<BatchConfig>;

  constructor(config: BatchConfig) {
    this.config = {
      concurrency: config.concurrency || 3,
      timeout: config.timeout || 30000,
      outputFormat: config.outputFormat || 'sharegpt',
      continueOnError: config.continueOnError ?? true,
    };
  }

  /**
   * 并行运行批处理
   * @param prompts - 待处理的 prompt 列表
   * @param executor - 单个 prompt 的执行函数
   * @returns 所有 prompt 的执行结果
   */
  async run(
    prompts: BatchPrompt[],
    executor: (prompt: BatchPrompt) => Promise<BatchItemResult>
  ): Promise<BatchItemResult[]> {
    if (!Array.isArray(prompts) || prompts.length === 0) {
      return [];
    }

    const results: BatchItemResult[] = [];
    const queue = [...prompts];
    let running = 0;

    Logger.info(
      `批处理启动: ${prompts.length} 个任务, 并发数 ${this.config.concurrency}`,
      'BatchProcessor'
    );

    return new Promise((resolve) => {
      const tryNext = () => {
        if (queue.length === 0 && running === 0) {
          const successCount = results.filter((r) => r.success).length;
          Logger.info(
            `批处理完成: ${successCount}/${results.length} 成功`,
            'BatchProcessor'
          );
          resolve(results);
          return;
        }

        while (running < this.config.concurrency && queue.length > 0) {
          const prompt = queue.shift()!;
          running++;

          // 带超时的执行，超时后清理定时器
          let timeoutId: ReturnType<typeof setTimeout> | undefined;
          const timeoutPromise = new Promise<BatchItemResult>((_, reject) => {
            timeoutId = setTimeout(
              () => reject(new Error('执行超时')),
              this.config.timeout
            );
          });

          const executeWithTimeout = Promise.race([
            executor(prompt).finally(() => {
              if (timeoutId !== undefined) clearTimeout(timeoutId);
            }),
            timeoutPromise,
          ]);

          executeWithTimeout
            .then((result) => {
              results.push(result);
            })
            .catch((err) => {
              if (timeoutId !== undefined) clearTimeout(timeoutId);
              const errorMessage = (err as Error).message;
              Logger.warn(
                `任务 ${prompt.id} 失败: ${errorMessage}`,
                'BatchProcessor'
              );
              results.push({
                id: prompt.id,
                response: '',
                success: false,
                duration: 0,
                error: errorMessage,
              });
            })
            .finally(() => {
              running--;
              tryNext();
            });
        }
      };

      tryNext();
    });
  }

  /**
   * 转换为 ShareGPT 格式
   * @param results - 批处理结果列表
   * @returns ShareGPT 对话格式数据
   */
  toShareGPT(results: BatchItemResult[]): ShareGPTConversation {
    return {
      conversations: results.flatMap((r) => [
        { from: 'human' as const, value: r.id },
        { from: 'gpt' as const, value: r.response },
      ]),
    };
  }

  /**
   * 转换为 JSONL 格式
   * @param results - 批处理结果列表
   * @returns JSONL 格式字符串
   */
  toJSONL(results: BatchItemResult[]): string {
    return results
      .map((r) =>
        JSON.stringify({ id: r.id, response: r.response, success: r.success })
      )
      .join('\n');
  }

  /**
   * 获取配置
   * @returns 只读的完整配置
   */
  getConfig(): Readonly<Required<BatchConfig>> {
    return { ...this.config };
  }
}

File: python/agent/loop/batch_processor.py
from __future__ import annotations

import asyncio
import json
from dataclasses import dataclass, field
from typing import Any

from agent.core.logger import StructuredLogger

log = StructuredLogger("batch_processor")


@dataclass
class BatchConfig:
    """批处理配置。

    Attributes:
        concurrency: 并发数。
        timeout: 单个任务超时（秒）。
        output_format: 输出格式（sharegpt/jsonl）。
        continue_on_error: 遇到错误是否继续处理下一个。
    """

    concurrency: int = 3
    timeout: float = 30.0
    output_format: str = "sharegpt"
    continue_on_error: bool = True


@dataclass
class BatchPrompt:
    """批处理提示条。

    Attributes:
        id: 唯一标识。
        text: 提示文本。
        system_prompt: 系统提示。
        metadata: 附加元数据。
    """

    id: str
    text: str
    system_prompt: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class BatchItemResult:
    """批处理单个任务结果。

    Attributes:
        id: 任务ID。
        response: 响应文本。
        success: 是否成功。
        duration: 耗时（秒）。
        error: 错误信息。
        metadata: 附加元数据。
    """

    id: str
    response: str = ""
    success: bool = False
    duration: float = 0.0
    error: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class ShareGPTConversation:
    """ShareGPT格式对话。

    Attributes:
        conversations: 对话轮次列表，每轮包含from/value。
    """

    conversations: list[dict[str, str]] = field(default_factory=list)


class BatchProcessor:
    """批处理引擎——并行执行任务，支持超时控制和错误处理。

    使用信号量控制并发数，支持ShareGPT和JSONL两种输出格式。

    Usage:
        config = BatchConfig(concurrency=5, timeout=30, output_format="jsonl")
        processor = BatchProcessor(config)
        prompts = [BatchPrompt(id="1", text="..."), BatchPrompt(id="2", text="...")]
        results = await processor.run(prompts, my_executor)
    """
    def __init__(self, config: BatchConfig | None = None) -> None:
        self._config = config or BatchConfig()

    @property
    def config(self) -> BatchConfig:
        return self._config

    async def run(
        self,
        prompts: list[BatchPrompt],
        executor,
    ) -> list[BatchItemResult]:
        if not prompts:
            return []

        results: list[BatchItemResult] = []
        queue = list(prompts)
        running = 0
        done_event = asyncio.Event()

        log.info(f"批处理启动: {len(prompts)} 个任务, 并发数 {self._config.concurrency}")

        async def try_next() -> None:
            nonlocal running

            if not queue and running == 0:
                success_count = sum(1 for r in results if r.success)
                log.info(f"批处理完成: {success_count}/{len(results)} 成功")
                done_event.set()
                return

            while running < self._config.concurrency and queue:
                prompt = queue.pop(0)
                running += 1

                async def execute_one(p: BatchPrompt) -> None:
                    nonlocal running
                    import time
                    start = time.monotonic()
                    try:
                        result = await asyncio.wait_for(
                            executor(p),
                            timeout=self._config.timeout,
                        )
                        result.duration = time.monotonic() - start
                        results.append(result)
                    except asyncio.TimeoutError:
                        results.append(BatchItemResult(
                            id=p.id,
                            success=False,
                            error="执行超时",
                            duration=self._config.timeout,
                        ))
                    except Exception as err:
                        results.append(BatchItemResult(
                            id=p.id,
                            success=False,
                            error=str(err),
                            duration=time.monotonic() - start,
                        ))
                    finally:
                        running -= 1
                        await try_next()

                asyncio.create_task(execute_one(prompt))

        await try_next()
        await done_event.wait()
        return results

    def to_sharegpt(self, results: list[BatchItemResult]) -> ShareGPTConversation:
        conversations: list[dict[str, str]] = []
        for r in results:
            conversations.append({"from": "human", "value": r.id})
            conversations.append({"from": "gpt", "value": r.response})
        return ShareGPTConversation(conversations=conversations)

    def to_jsonl(self, results: list[BatchItemResult]) -> str:
        lines = []
        for r in results:
            lines.append(json.dumps({
                "id": r.id,
                "response": r.response,
                "success": r.success,
            }, ensure_ascii=False))
        return "\n".join(lines)

    def get_config(self) -> BatchConfig:
        return self._config
